crates/mev-share-rpc-api/src/auth.rs

```diff
@@ -0,0 +1,148 @@
+//! A layer responsible for implementing flashbots-style authentication
+//! by signing the request body with a private key and adding the signature
+//! to the request headers.
+
+use std::{
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use ethers_core::{types::H256, utils::keccak256};
+use ethers_signers::Signer;
+use futures_util::future::BoxFuture;
+
+use http::{header::HeaderValue, HeaderName, Request};
+use hyper::Body;
+
+use tower::{Layer, Service};
+
+/// Layer that applies [`FlashbotsSigner`] which adds a request header with a signed payload.
+#[derive(Clone)]
+pub struct FlashbotsSignerLayer<S> {
+    signer: Arc<S>,
+}
+
+impl<S> FlashbotsSignerLayer<S> {
+    /// Creates a new [`FlashbotsSignerLayer`] with the given signer.
+    pub fn new(signer: Arc<S>) -> Self {
+        FlashbotsSignerLayer { signer }
+    }
+}
+
+impl<S: Clone, I> Layer<I> for FlashbotsSignerLayer<S> {
+    type Service = FlashbotsSigner<S, I>;
+
+    fn layer(&self, inner: I) -> Self::Service {
+        FlashbotsSigner {
+            signer: self.signer.clone(),
+            inner,
+        }
+    }
+}
+
+/// Middleware that adds a request header with a signed payload.
+/// For more info, see https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint#authentication
+#[derive(Clone)]
+pub struct FlashbotsSigner<S, I> {
+    signer: Arc<S>,
```

Comment:
I think we can remove the Arc here and instead require that the Signer is clone:

> S: Signer + Clone + Send + 'static,

Reply:
fixed 👍